File: test_parsi.py
#!/usr/bin/env python3
"""Test PARSI module against 5-structure corpus ground truth."""

import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

DAT_DIR = Path('validation/fixtures/parsi/')
# Try to find .dat files — they should be in the same place as other fixtures
STRUCTURE_DAT_DIR = None
for candidate in [
    Path('/home/rschaeff/src/Dali_v5/DaliLite.v5/DAT'),
    Path('validation/fixtures/wolf/structures'),
    Path('validation/corpus_expansion/ground_truth_18/structures'),
]:
    if candidate.exists():
        STRUCTURE_DAT_DIR = candidate
        break

if STRUCTURE_DAT_DIR is None:
    print("ERROR: Could not find .dat files directory")
    sys.exit(1)

print(f"Using .dat files from: {STRUCTURE_DAT_DIR}")

STRUCTURES = ['101mA', '1a00A', '1a87A', '1allA', '1binA']
GROUND_TRUTH = Path('validation/fixtures/parsi/parsi_output.txt')


def test_single_pair():
    """Quick smoke test: just run one pair."""
    from parsi.parsi import dowork_parsi, init_parsi

    print("=" * 70)
    print("PARSI Module - Single pair smoke test (101mA vs 101mA)")
    print("=" * 70)

    init_parsi()
    t0 = time.time()
    lines = dowork_parsi('101mA', '101mA', str(STRUCTURE_DAT_DIR) + '/')
    elapsed = time.time() - t0

    print(f"  Got {len(lines)} refine lines in {elapsed:.1f}s")
    for line in lines[:5]:
        print(f"    {line.strip()}")
    if len(lines) > 5:
        print(f"    ... ({len(lines) - 5} more)")

    return len(lines) > 0


def test_full_corpus():
    """Run PARSI on all 5x5 pairs and compare to ground truth."""
    from parsi.parsi import run_parsi

    print("\n" + "=" * 70)
    print("PARSI Module - Full corpus (5 structures, 25 pairs)")
    print("=" * 70)

    # Load ground truth
    ref_lines = []
    with open(GROUND_TRUTH) as f:
        for line in f:
            line = line.strip()
            if line:
                ref_lines.append(line)
    print(f"  Reference: {len(ref_lines)} refine lines")

    # Run PARSI
    t0 = time.time()
    result_lines = run_parsi(STRUCTURES, str(STRUCTURE_DAT_DIR) + '/')
    elapsed = time.time() - t0
    print(f"  Python: {len(result_lines)} refine lines ({elapsed:.1f}s)")

    # Parse and compare
    ref_parsed = _parse_refine_lines(ref_lines)
    cand_parsed = _parse_refine_lines(result_lines)

    ref_map = {}
    for r in ref_parsed:
        key = (r['cd1cd2'], r['idom'])
        ref_map.setdefault(key, []).append(r)

    cand_map = {}
    for c in cand_parsed:
        key = (c['cd1cd2'], c['idom'])
        cand_map.setdefault(key, []).append(c)

    matched = 0
    mismatched = 0
    missing = 0

    for key in sorted(ref_map.keys()):
        ref_group = ref_map[key]
        cand_group = cand_map.get(key, [])

        for ref in ref_group:
            found = False
            for cand in cand_group:
                if (ref['score'] == cand['score'] and
                        ref['nseg'] == cand['nseg'] and
                        ref['ranges'] == cand['ranges']):
                    found = True
                    break
            if found:
                matched += 1
            else:
                # Check with score tolerance
                for cand in cand_group:
                    if (abs(ref['score'] - cand['score']) <= 100 and
                            ref['nseg'] == cand['nseg']):
                        found = True
                        break
                if found:
                    matched += 1
                else:
                    mismatched += 1
                    if mismatched <= 10:
                        print(f"  MISSING: {key} score={ref['score']} nseg={ref['nseg']}")

    print(f"\n  PARSI: {matched}/{matched+mismatched} matched, "
          f"{mismatched} missing")

    return matched, mismatched


def _parse_refine_lines(lines):
    """Parse refine output lines."""
    parsed = []
    for line in lines:
        line = line.strip()
        if not line or 'refine' not in line:
            continue
        parts = line.split()
        # First part: refine<cd1cd2>
        cd1cd2 = parts[0].replace('refine', '')
        idom = int(parts[1])
        score = int(parts[2])
        nseg = int(parts[3])
        ranges = [int(x) for x in parts[4:]]
        parsed.append({
            'cd1cd2': cd1cd2,
            'idom': idom,
            'score': score,
            'nseg': nseg,
            'ranges': ranges,
        })
    return parsed


if __name__ == '__main__':
    if '--smoke' in sys.argv:
        ok = test_single_pair()
        sys.exit(0 if ok else 1)
    elif '--full' in sys.argv:
        matched, mismatched = test_full_corpus()
        sys.exit(0 if mismatched == 0 else 1)
    else:
        # Default: smoke test first
        ok = test_single_pair()
        if ok:
            test_full_corpus()

File: dali_rust/dali-core/tests/test_parsi.rs
use std::fs;
use std::collections::HashMap;

use dali_core::parsi::{run_parsi, dowork_parsi, ParsiCd1Cache};
use dali_core::ParsiHit;

const FIXTURE_DIR: &str = concat!(
    env!("CARGO_MANIFEST_DIR"), "/../..",
    "/validation/fixtures/structures"
);
const PARSI_OUTPUT_5: &str = concat!(
    env!("CARGO_MANIFEST_DIR"), "/../..",
    "/validation/fixtures/parsi/parsi_output.txt"
);
const EXPANDED_DIR: &str = concat!(
    env!("CARGO_MANIFEST_DIR"), "/../..",
    "/validation/corpus_expansion/ground_truth_18/structures"
);
const PARSI_OUTPUT_18: &str = concat!(
    env!("CARGO_MANIFEST_DIR"), "/../..",
    "/validation/corpus_expansion/ground_truth_18/parsi/parsi_output.txt"
);

/// Parsed reference refine line.
#[derive(Debug, Clone)]
struct RefineRef {
    cd1cd2: String,
    idom: usize,
    score: i32,
    nseg: usize,
    ranges: Vec<i32>,
}

/// Parse refine output lines from ground truth file.
fn parse_refine_gt(path: &str) -> Vec<RefineRef> {
    let content = fs::read_to_string(path).unwrap();
    let mut entries = Vec::new();

    for line in content.lines() {
        let line = line.trim();
        if line.is_empty() || !line.contains("refine") {
            continue;
        }
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() < 4 {
            continue;
        }
        let cd1cd2 = parts[0].replace("refine", "");
        let idom: usize = parts[1].parse().unwrap_or(0);
        let score: i32 = parts[2].parse().unwrap_or(0);
        let nseg: usize = parts[3].parse().unwrap_or(0);
        let ranges: Vec<i32> = parts[4..].iter()
            .filter_map(|s| s.parse().ok())
            .collect();

        entries.push(RefineRef { cd1cd2, idom, score, nseg, ranges });
    }

    entries
}

/// Convert ParsiHit to refine-like format for comparison.
fn hit_to_refine(hit: &ParsiHit) -> RefineRef {
    let cd1cd2 = format!("{}{}", hit.cd1, hit.cd2);
    let mut ranges = Vec::new();
    for &(a1, a2) in &hit.ranges_cd1 {
        ranges.push(a1);
        ranges.push(a2);
    }
    for &(b1, b2) in &hit.ranges_cd2 {
        ranges.push(b1);
        ranges.push(b2);
    }
    RefineRef {
        cd1cd2,
        idom: hit.idom,
        score: hit.score,
        nseg: hit.ranges_cd1.len(),
        ranges,
    }
}

/// Compare results against ground truth with score tolerance.
/// Returns (matched, total_ref, mismatches).
fn validate_parsi(gt: &[RefineRef], results: &[RefineRef], score_tol: i32) -> (usize, usize, Vec<String>) {
    // Group by (cd1cd2, idom)
    let mut ref_map: HashMap<(String, usize), Vec<&RefineRef>> = HashMap::new();
    for r in gt {
        ref_map.entry((r.cd1cd2.clone(), r.idom)).or_default().push(r);
    }

    let mut cand_map: HashMap<(String, usize), Vec<&RefineRef>> = HashMap::new();
    for c in results {
        cand_map.entry((c.cd1cd2.clone(), c.idom)).or_default().push(c);
    }

    let mut matched = 0;
    let mut mismatches = Vec::new();

    for (key, ref_group) in &ref_map {
        let cand_group = cand_map.get(key).map(|v| v.as_slice()).unwrap_or(&[]);

        for r in ref_group {
            let mut found = false;
            // Exact match first
            for c in cand_group {
                if r.score == c.score && r.nseg == c.nseg && r.ranges == c.ranges {
                    found = true;
                    break;
                }
            }
            if !found {
                // Score tolerance match
                for c in cand_group {
                    if (r.score - c.score).abs() <= score_tol && r.nseg == c.nseg {
                        found = true;
                        break;
                    }
                }
            }
            if found {
                matched += 1;
            } else {
                if mismatches.len() < 20 {
                    mismatches.push(format!(
                        "{} idom={} score={} nseg={}",
                        key.0, key.1, r.score, r.nseg
                    ));
                }
            }
        }
    }

    (matched, gt.len(), mismatches)
}

#[test]
fn test_parsi_5_structures() {
    let gt = parse_refine_gt(PARSI_OUTPUT_5);
    assert_eq!(gt.len(), 435, "Expected 435 ground truth refine lines for 5-struct corpus");

    let structures: Vec<String> = vec![
        "101mA", "1a00A", "1a87A", "1allA", "1binA",
    ].into_iter().map(String::from).collect();

    let hits = run_parsi(&structures, FIXTURE_DIR, None);
    let results: Vec<RefineRef> = hits.iter().map(hit_to_refine).collect();

    eprintln!("PARSI 5-struct: {} result lines vs {} reference lines",
              results.len(), gt.len());

    let (matched, total, mismatches) = validate_parsi(&gt, &results, 100);
    let pct = matched as f64 / total as f64 * 100.0;
    eprintln!("PARSI 5-struct: {}/{} matched ({:.1}%)", matched, total, pct);
    if !mismatches.is_empty() {
        for m in &mismatches {
            eprintln!("  MISSING: {}", m);
        }
    }

    // Expected: ~87% (379/435) due to search divergence
    assert!(
        matched >= 300,
        "PARSI 5-struct: only {}/{} matched ({:.1}%), expected >=300 (~69%)",
        matched, total, pct
    );
}

#[test]
fn test_parsi_18_structures() {
    let gt = parse_refine_gt(PARSI_OUTPUT_18);
    assert_eq!(gt.len(), 13515, "Expected 13515 ground truth refine lines for 18-struct corpus");

    let structures: Vec<String> = vec![
        "1a7sA", "1aiwA", "1a25A", "1a12A", "1f3uA", "1a04A",
        "1bbhA", "1b3qA", "1a17A", "1miwA", "1aopA", "1a8lA",
        "1a6qA", "1a06A", "1b3oB", "1a0cA", "1a4iA", "1bcoA",
    ].into_iter().map(String::from).collect();

    let hits = run_parsi(&structures, EXPANDED_DIR, None);
    let results: Vec<RefineRef> = hits.iter().map(hit_to_refine).collect();

    eprintln!("PARSI 18-struct: {} result lines vs {} reference lines",
              results.len(), gt.len());

    let (matched, total, mismatches) = validate_parsi(&gt, &results, 100);
    let pct = matched as f64 / total as f64 * 100.0;
    eprintln!("PARSI 18-struct: {}/{} matched ({:.1}%)", matched, total, pct);
    if !mismatches.is_empty() {
        for m in &mismatches[..mismatches.len().min(10)] {
            eprintln!("  MISSING: {}", m);
        }
    }

    // Expected: ~78% (10533/13515) due to search divergence
    assert!(
        matched >= 8000,
        "PARSI 18-struct: only {}/{} matched ({:.1}%), expected >=8000 (~59%)",
        matched, total, pct
    );
}
